docs/RFCS/20220602_fine_grained_cpu_attribution.md

```diff
@@ -0,0 +1,615 @@
+- Feature Name: Fine-grained CPU attribution
+- Status: accepted
+- Start Date: 2022-06-02
+- Authors: Irfan Sharif, Austen McClernon
+- RFC PR: [#82356](https://github.com/cockroachdb/cockroach/pull/82356)
+- Cockroach Issue: N/A
+
+# Summary
+
+We propose using a patched Go runtime to track CPU use at the level of
+individual goroutines, using the primitive to derive an accurate resource
+consumption signal in various CRDB subsystems. The compiler, language, and
+tooling will remain the same.
+
+*NOTE: This RFC started off as an internal gdoc (grep for "Fine-grained CPU
+attribution" with valuable commentary and additional links).*
+
+## Motivation
+
+CRDB lacks primitives to attribute CPU usage to specific scopes (ranges,
+tenants, sessions, statements) or processes (pebble compactions, snapshot
+generation). For subsystems that rely on such attribution, we use proxy signals
+like '# of batch requests' or 'size of read/write request' as observed by a
+range/replica or a given tenant. With respect to modeling CPU usage, this is:
+- Inaccurate: two proportions of the signal can translate to different
+  proportions of actual CPU usage;
+- Imprecise: repeated equal measurements of the signal can correspond to varied
+  actual CPU usage.
+
+The inaccuracy and imprecision are due to:
+- Different batch requests having different compositions with respect to the
```

Comment:
nit: All of these relate to the request types and their CPU usage - however we could approximate them with a model if these variables were the only variables. One thing is missing, is the same exact batch request on two stores may have different CPU usage profiles depending on sublevels, contention etc.